packages/functional-tests/src/tests/actor-attachment-test.ts

```diff
@@ -0,0 +1,175 @@
+import { Box } from '@microsoft/gltf-gen';
+import { create } from 'domain';
+
+
+
+
+export default class ActorAttachmentTest extends Test {
+	public expectedResultDescription = "Actors attaching to avatar points";
+	private assets: MRE.AssetContainer;
+
+	//Limited subset of AttachmentPoints for testing
+	private attachments: AttachPoint[] =[ 
+	 'head'
+	, 'neck'
+	, 'hips'
+	, 'left-eye'
+	, 'left-foot'
+	, 'left-hand'
+	, 'right-eye'
+	, 'right-foot'
+	, 'right-hand'
+	];
+
+	//Incremented by clicking grey cube
+	private attachmentIndex = 0;
+
+	//Oblong box that gets attached to the avatar of the user running this test
+	private attachedCube : MRE.Actor;
+
+	//Reference to user that started this test, this user will receive attachments
+	private rootActor: MRE.Actor;
+
+	//Switch to determine if attachments are fully destroyed and remade in between changing attachment points
+	//The state is displayed by the color of the cube to the left of the attachment cycle button
+	//True = Red Cube  = attachedCube is destroyed and re-created
+	//False = Blue Cube = attachedCube is detached and attached without being destroyed
+	//Currently these produce different results in the end transform of attachedCube
+	private reCreateCubeTest = true;
+
+	//Destroy if necessary and create oblong attachment cube
+	private createCubeAttachment() {
+		if(this.attachedCube)
+			this.attachedCube.destroy();
+		this.attachedCube = MRE.Actor.Create(this.app.context, {
+			actor: {
+				name: 'cube1',
+				parentId: this.rootActor.id,
+				appearance: {
+					meshId: this.assets.createBoxMesh('smallBox', 1, 1, 1).id
+				},
+				transform: {
+					local: {
+						scale: { x: 2, y: 0.5, z: 0.5 }
+					}
+				}
+			
+			}
+		});
+	}
+
+	//Materials for displaying state of reCreateCubeTest variable
+	private redMat: MRE.Material;
+	private blueMat: MRE.Material;
+
+	public async run(root: MRE.Actor): Promise<boolean> {
+		this.rootActor = root;
+		this.assets = new MRE.AssetContainer(this.app.context);
+		const deg45 = MRE.Quaternion.FromEulerAngles(0, 0, -Math.PI / 4);
+
+		//Create Materials and actors
+		this.redMat = this.assets.createMaterial('redBall', {
+			color: MRE.Color3.Red()
+		});
+		this.blueMat = this.assets.createMaterial('blueBall', {
+			color: MRE.Color3.Blue()
+		});
+
+		const buttonCube = MRE.Actor.Create(this.app.context, {
+			actor: {
+				name: 'cube1',
+				parentId: this.rootActor.id,
+				appearance: {
+					meshId: this.assets.createBoxMesh('smallBox', 1, 1, 1).id
+				},
+				collider: { geometry: { shape: MRE.ColliderType.Box } },
+				transform: {
+					local: {
+						position: { x: 0, y: .5 },
+
+						scale: { x: 0.5, y: 0.5, z: 0.5 }
+					}
+				}
+			}
+		});
+		
+		const CreateCube = MRE.Actor.Create(this.app.context, {
+			actor: {
+				name: 'cube1',
+				parentId: this.rootActor.id,
+				appearance: {
+					meshId: this.assets.createBoxMesh('smallBox', 1, 1, 1).id,
```

Comment:
>this.assets.createBoxMesh('smallBox', 1, 1, 1) [](start = 13, length = 46)

There are a couple duplicates of this "small box" mesh that could be merged into a single reused instance. One pattern I've used before is get-or-create, where you check if there's a mesh by that name in the asset container already. Saves you the trouble of passing the reference around.

Reply:
Could you provide an example of how that pattern could be integrated into this code? Would using a conditional operator in-line to check for the existence make sense or do we want to provide a different example?

Reply:
Let me know what you think of the latest iteration

Reply:
`const boxMesh = this.assets.meshes.find(m => m.name === "smallBox") ?? this.assets.createBoxMesh("smallBox", 1,1,1);` is how I usually do it. This way works too though.

---
In reply to: [445098060](https://github.com/microsoft/mixed-reality-extension-sdk/pull/631#discussion_r445098060) [](ancestors = 445098060)